packages/next/src/server/app-render/app-render.tsx

```diff
@@ -77,6 +75,18 @@ import { PAGE_SEGMENT_KEY } from '../../shared/lib/constants'
 import { NEXT_DYNAMIC_NO_SSR_CODE } from '../../shared/lib/lazy-dynamic/no-ssr-error'
 import { warn } from '../../build/output/log'
 
+// Import builtin react directly to avoid require cache conflicts
+let React: typeof import('next/dist/compiled/react')
+let ReactDOMServer: typeof import('next/dist/compiled/react-dom/server.browser')
+
+if (process.env.NEXT_PREBUNDLED_REACT === 'experimental') {
```

Comment:
Do we think we can wrap this prat as a internal lib, and import react packages from there?

Reply:
Ideally, branching won't be needed if the require hook is working correctly. So it's more necessary to setup `NEXT_PREBUNDLED_REACT` and the require hook early, and that kinda feels wrong. Will do something for that.

Some of the require hook related fixes are blocked by #47857, so I won't expect it to be solved elegantly any time soon this week.